client/state-db/src/lib.rs

```diff
@@ -339,30 +342,39 @@ impl<BlockHash: Hash + MallocSizeOf, Key: Hash + MallocSizeOf> StateDbSync<Block
 		if let Some(ref mut pruning) = self.pruning {
 			pruning.note_canonical(hash, &mut commit);
 		}
-		self.prune(&mut commit);
+		self.prune(&mut commit)?;
 		Ok(commit)
 	}
 
 	fn best_canonical(&self) -> Option<u64> {
 		self.non_canonical.last_canonicalized_block_number()
 	}
 
-	fn is_pruned(&self, hash: &BlockHash, number: u64) -> bool {
-		match self.mode {
+	fn is_pruned(&self, hash: &BlockHash, number: u64) -> IsPruned {
+		let is_pruned = match self.mode {
 			PruningMode::ArchiveAll => false,
```

Comment:
This should result in a` Maybe`

Reply:
Changing to `Maybe` will violate the original logic, since `ArchiveAll`  mode will not prune any state

Reply:
The issue is again with unknow block hashes. Original logic was a bit incorrect here too, that's why there's a special case here: https://github.com/paritytech/substrate/blob/d0c7710a3b4a76f072101819ac4db728b27e0b96/client/db/src/lib.rs#L2287 
That special handling for archive mode could be removed.

On the other hand `is_pruned` currently answers the question "was this block hash+number hash ever deleted from the database". So techincally the answer for archive mode should be "no" indeed.